test/run.d

```diff
@@ -0,0 +1,342 @@
+
+    // allow overwrites from the environment
+    resultsDir = environment.get("RESULTS_DIR", resultsDir);
+    hostDMD = environment.get("HOST_DMD", "dmd");
+
+    // bootstrap all needed environment variables
+    auto env = getEnvironment;
+
+    // default target
+    if (!args.length)
+        args = ["all"];
+
+    alias normalizeTestName = f => format!"%s/%s"(f.absolutePath.dirName.baseName, f.baseName);
+    auto targets = args
+        .predefinedTargets // preprocess
+        .map!normalizeTestName
+        .array
+        .filterTargets;
+
+    if (targets.length > 0)
+    {
+        if (verbose)
+        {
+            log("================================================================================");
+            foreach (key, value; env)
+                log("%s=%s", key, value);
+            log("================================================================================");
+        }
+        auto taskPool = new TaskPool(jobs);
+        scope(exit) taskPool.finish();
+        ensureToolsExists;
+        foreach (target; taskPool.parallel(targets, 1))
+        {
+            auto args = [resultsDir.buildPath("d_do_test"), target];
+            log("run: %-(%s %)", args);
+            spawnProcess(args, env, Config.none, scriptDir).wait;
+        }
+    }
+}
+
+/**
+Builds the binary of the tools required by the testsuite.
+Does nothing if the tools already exist and are newer than their source.
+*/
+void ensureToolsExists()
+{
+    static toolsDir = scriptDir.buildPath("tools");
+    resultsDir.mkdirRecurse;
+    auto tools = [
+        "d_do_test",
+        "sanitize_json",
+    ];
+    foreach (tool; tools.parallel(1))
+    {
+        auto targetBin = resultsDir.buildPath(tool);
+        auto sourceFile = toolsDir.buildPath(tool ~ ".d");
+        if (targetBin.timeLastModified.ifThrown(SysTime.init) > sourceFile.timeLastModified)
+            writefln("%s is already up-to-date", tool);
+        else
+        {
+            auto command = [hostDMD, "-of"~targetBin, sourceFile];
+            writefln("Executing: %-(%s %)", command);
+            spawnProcess(command).wait;
+        }
+    }
+
+    // ensure output directories exist
+    foreach (dir; testDirs)
+        resultsDir.buildPath(dir).mkdirRecurse;
+}
+
+/**
+Goes through the target list and replaces short-hand targets with their expanded version.
+Special targets:
+- clean -> removes resultsDir + immediately stops the runner
+*/
+auto predefinedTargets(string[] targets)
+{
+    static findFiles(string dir)
+    {
+        return scriptDir.buildPath(dir).dirEntries("*{.d,.sh}", SpanMode.shallow).map!(e => e.name);
+    }
+
+    Appender!(string[]) newTargets;
+    foreach (t; targets)
+    {
+        t = t.buildNormalizedPath; // remove trailing slashes
+        switch (t)
+        {
+            case "clean":
+                resultsDir.rmdirRecurse;
+                exit(0);
+                break;
+
+            case "run_runnable_tests", "runnable":
```

Comment:
The "run_" prefix seems redundant, but I'm just being picky.

Reply:
`run_runnable_tests` is the name of the existing Makefile target and at least for the transition I wanted to keep the most commonly used Makefile targets